### PR TITLE
TVM CLI to compile models and generate config file

### PR DESCRIPTION
**Pull request by LucaFos**
Python script to compile a model using TVM and generate a C++ header
file to be used in the inference engine. Support .onnx/.pb models.
The tool takes the model zoo definition.yaml file as an input

A dockerfile is provided to build an environment with all required
dependencies to run the cli tool and build the example pipeline for
yolo_v2_tiny. The upstream TVM docker images are not maintained very
well; hence we manually build and install tvm to be able to use a newer
version and more reproducible environment.

Change-Id: Ie74f9aef4ebacd8a9f446ea88f6e8443be6238ec
Signed-off-by: Luca Foschiani <luca.foschiani@arm.com>
Issue-Id: SCM-874

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -148,17 +148,15 @@ they want to be able to refer to the model, and know that that reference will
...
-1. Verbose description of how the training data is obtained and pre-processed
-1. Verbose discussion of bias and limitation in the training data
-1. Guide on how to do transfer learning using this model
-1. Share performance metrics on inference hardware
-1. Share accuracy metrics on well known datasets.
+In this model zoo, a CLI tool has been provided to compile models using TVM.
+This forms part of the validation workflow of the zoo as well as model
+deployment workflow for Autoware.
 
-Do commit any scripts or code you can share in the same folder as the models.
```
Comment (JWhitleyWork) [inline README.md:164]: Why was this removed in favor of the TVM CLI section rather than just adding a new section?

File: scripts/tvm_cli/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,17 @@
+FROM ubuntu:18.04
+
+COPY install_tvm.sh /tmp/install_tvm.sh
+RUN /tmp/install_tvm.sh
+
+RUN pip3 install --upgrade pip && pip3 install onnx tensorflow pyyaml jinja2
```
Comment (JWhitleyWork) [inline scripts/tvm_cli/Dockerfile:6]: When running this command in the Docker build, I get the following error:

```
ERROR: After October 2020 you may experience errors when installing or updating packages. This is because pip will change the way that it resolves dependency conflicts.

We recommend you use --use-feature=2020-resolver to test your packages with the new resolver before it becomes the default.

google-auth 1.22.1 requires setuptools>=40.3.0, but you'll have setuptools 39.0.1 which is incompatible.
tensorboard 2.3.0 requires setuptools>=41.0.0, but you'll have setuptools 39.0.1 which is incompatible.
tensorflow 2.3.1 requires numpy<1.19.0,>=1.16.0, but you'll have numpy 1.19.2 which is incompatible.
```

It does not stop the build but I'm worried about the deprecation warning causing errors very soon. Thoughts?

Comment (JWhitleyWork) [reply scripts/tvm_cli/Dockerfile:6]: Additionally, the `setuptools` errors look problematic too.

File: scripts/tvm_cli/README.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+               scripts/tvm_cli
+```
+
+The CLI can now be invoked as a container
+
+```bash
+$ docker run -it --rm -v `pwd`:`pwd` -w `pwd` \
+    -u $(id -u ${USER}):$(id -g ${USER}) \
+    autoware-model-zoo/tvm_cli:local -h
+```
```
Comment (JWhitleyWork) [inline scripts/tvm_cli/README.md:24]: When running this as a test, I get the following output at the top:

```
2020-10-08 15:18:00.785374: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory
2020-10-08 15:18:00.785405: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
```

I added `--gpus all` since I have both an Nvidia graphics card and the `nvidia-container-toolkit` installed. It seems like this would also require CUDA to be installed in the container to support Nvidia GPUs (and maybe ROCm for AMD GPUs). Do we want to add these to the container?

**Review by JWhitleyWork — CHANGES_REQUESTED** [review]
See comments.

File: README.md (hunk 1 above)
Comment (LucaFos) [reply README.md:164]: This line is a duplicate of https://github.com/autowarefoundation/modelzoo/commit/1899638d956244649a5782ee8b64a873b4ea529a#diff-04c6e90faac2675aa89e2176d2eec7d8R133.

File: scripts/tvm_cli/Dockerfile (hunk 2 above)
Comment (LucaFos) [reply scripts/tvm_cli/Dockerfile:6]: Thanks for pointing these out, I was able to reproduce the first issue but I'm not getting any error/warning regarding setuptools. Was the setuptools error coming from the setuptools installation?

File: scripts/tvm_cli/README.md (hunk 3 above)
Comment (LucaFos) [reply scripts/tvm_cli/README.md:24]: I'm currently working on a subsequent patch, part of this patch is adding CUDA support to the TVM CLI. In my opinion we should leave the current patch as-is (w.r.t CUDA) because we're adding CUDA on a future patch anyway.
Please let me know what are your thoughts about this.

File: scripts/tvm_cli/Dockerfile (hunk 2 above)
Comment (LiyouZhou) [reply scripts/tvm_cli/Dockerfile:6]: @lucafos
I think @JWhitleyWork meant
```
google-auth 1.22.1 requires setuptools>=40.3.0, but you'll have setuptools 39.0.1 which is incompatible.
tensorboard 2.3.0 requires setuptools>=41.0.0, but you'll have setuptools 39.0.1 which is incompatible.
tensorflow 2.3.1 requires numpy<1.19.0,>=1.16.0, but you'll have numpy 1.19.2 which is incompatible.
```

Comment (LiyouZhou) [reply scripts/tvm_cli/Dockerfile:6]: @LucaFos, you may need to update setuptools to the latest version before pip install anything.

Comment (LucaFos) [reply scripts/tvm_cli/Dockerfile:6]: Just checked, the setuptools issue got solved together with the deprecation warning.

Comment (JWhitleyWork) [reply scripts/tvm_cli/Dockerfile]: Verified as resolved. Thanks!

**Review by JWhitleyWork — APPROVED** [review]
LGTM
